Automatically run dynamo in torch.cond

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107932

The key modifications include:
1.  moving cond from functorch/experimental/ to torch/_higher_order_ops/cond.py
2. always turn on dynamo in cond_compiled. This requires skipping tests when run on a platform that's not dynamo
3. change test behaviors after turning on torch.compile  for many eager tests.
4. expected fail a few tests (make_fx (tracing_mode="symbolic") in test_control_flow.py due to SymBool. Current plan is to wait for https://github.com/pytorch/pytorch/pull/107850 and forward fix.
5. Add new tests in test_control_flow.py to check cond can use dynamo to automatically lift the closures as input.

> How did we get Dynamo to error out on torch.ops.higher_order.cond if there are any graph breaks? 

The current implementation is to catch the UserError.GRAPH_BREAK_IN_CONTROL_FLOW and once saw it, we directly raise: once in [break_graph_if_unsupported](https://github.com/pytorch/pytorch/blob/bad3f2db40ffa491181fc7df5dc310830326d9ec/torch/_dynamo/symbolic_convert.py#L1250), which catches and raises for call_function (entry point of higher order operator)  and a few others. The raised exception is caught and raised again in [step](https://github.com/pytorch/pytorch/blob/bad3f2db40ffa491181fc7df5dc310830326d9ec/torch/_dynamo/symbolic_convert.py#L691), where all instructions' exceptions are handled.


cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov

ydwu4 committed:
Automatically run on dynamo in torch.cond

[ghstack-poisoned]

ydwu4 committed:
Update on "Automatically run dynamo in torch.cond"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Automatically run dynamo in torch.cond"


The key modification is:
1.  moving cond from functorch/experimental/ to torch/_higher_order
2. always turn on dynamo in cond_compiled
3. change test behaviors after turning on torch.compile  for some eager run tests.
4. expected fail a few tests due to SymBool. Current plan is to wait for https://github.com/pytorch/pytorch/pull/107850 and forward fix.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

ydwu4 committed:
Update on "Automatically run dynamo in torch.cond"


The key modification is:
1.  moving cond from functorch/experimental/ to torch/_higher_order
2. always turn on dynamo in cond_compiled
3. change test behaviors after turning on torch.compile  for some eager run tests.
4. expected fail a few tests due to SymBool. Current plan is to wait for https://github.com/pytorch/pytorch/pull/107850 and forward fix.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

ydwu4 committed:
Update on "Automatically run dynamo in torch.cond"


The key modification is:
1.  moving cond from functorch/experimental/ to torch/_higher_order
2. always turn on dynamo in cond_compiled
3. change test behaviors after turning on torch.compile  for some eager run tests.
4. expected fail a few tests due to SymBool. Current plan is to wait for https://github.com/pytorch/pytorch/pull/107850 and forward fix.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

avikchaudhuri reviewed (approved):
Discussed changes offline.

avikchaudhuri commented on test/dynamo/test_export.py:
```diff
@@ -2842,8 +2842,8 @@ def f(x):
         with self.assertRaisesRegex(
-            torch._dynamo.exc.UserError,
-            "Expected 4 arguments",
+            torch._dynamo.exc.ArgsMismatchError,
```
Was this expected to be UserError in export DB?

ydwu4 replied:
Not yet.

avikchaudhuri commented on test/dynamo/test_export.py:
```diff
@@ -2869,7 +2869,7 @@ def f_non_list_operands(x):
         with self.assertRaisesRegex(
             torch._dynamo.exc.UserError,
-            "Expected a list but got",
+            "Expected a list/tuple but got",
```
list or tuple

avikchaudhuri commented on test/functorch/test_control_flow.py:
```diff
@@ -650,8 +654,8 @@ def f(x, y):
-            CondOpArgsMismatchError,
-            "Expected to return same number of outputs but got",
+            torch._dynamo.exc.UserError,
+            "Expected branch out type to be a single tensor",
```
`branch out type` isn't very clear. Maybe `Expected branch to return a single tensor`?

avikchaudhuri commented on test/functorch/test_control_flow.py:
```diff
@@ -286,13 +287,14 @@ def false_fn(x):
 
         def f(x):
-            return cond(x.all(), true_fn, false_fn, (x,))
+            return cond(x.all(), true_fn, false_fn, [x,])
```
revert

avikchaudhuri reviewed (commented):
nits

ydwu4 committed:
Update on "Automatically run dynamo in torch.cond"


The key modification is:
1.  moving cond from functorch/experimental/ to torch/_higher_order_ops/cond.py
2. always turn on dynamo in cond_compiled except on windows because torch.compile doesn't support windows rn.
3. change test behaviors after turning on torch.compile  for many eager tests.
4. expected fail a few tests in test_control_flow.py due to SymBool. Current plan is to wait for https://github.com/pytorch/pytorch/pull/107850 and forward fix.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

ydwu4 commented on functorch/experimental/control_flow.py:
```diff
@@ -1,2 +1,6 @@
-from ._cond import cond, UnsupportedAliasMutationException  # noqa: F401
+from torch._higher_order_ops.cond import (  # noqa: F401  # noqa: F401
+    cond_compiled as cond,
```
Yeah, I can add a top-level torch.cond.

gmagogsfm replied:
Yes, it is good to have functorch.experimental.control_flow to avoid excessive breakages.

will there be a PR for adding torch.cond?

ydwu4 replied:
I think the best case is to have the torch.cond landed after SymBool PR. But the SymBool one seems difficult to land today. It might be possible over the weekend.

ydwu4 committed:
Update on "Automatically run dynamo in torch.cond"


The key modification is:
1.  moving cond from functorch/experimental/ to torch/_higher_order_ops/cond.py
2. always turn on dynamo in cond_compiled except on windows because torch.compile doesn't support windows rn.
3. change test behaviors after turning on torch.compile  for many eager tests.
4. expected fail a few tests in test_control_flow.py due to SymBool. Current plan is to wait for https://github.com/pytorch/pytorch/pull/107850 and forward fix.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

ydwu4 committed:
Update on "Automatically run dynamo in torch.cond"


The key modification is:
1.  moving cond from functorch/experimental/ to torch/_higher_order_ops/cond.py
2. always turn on dynamo in cond_compiled except on windows because torch.compile doesn't support windows rn.
3. change test behaviors after turning on torch.compile  for many eager tests.
4. expected fail a few tests in test_control_flow.py due to SymBool. Current plan is to wait for https://github.com/pytorch/pytorch/pull/107850 and forward fix.
5. Add new tests in test_control_flow.py to check cond can use dynamo to automatically lift the closures as input.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -1038,12 +1038,17 @@ def false_fn(x):
         mod_for_compile = torch.compile(Foo(), backend=cnt, dynamic=True)
         mod_for_eager = Foo()
 
+        ref = torch.tensor(72.0)
```
Does this get used anywhere?

ydwu4 replied:
yeah, it's used to compare with mod_for_compile results in L1052

zou3519 reviewed (commented):
The largest item from me is: In eager-mode cond, cond should have the same behavior in eager-mode and under torch.compile. That means that if we are unable to torch.compile(fullgraph=True) it in, then it should error out.

This happens in eager-mode cond, but not cond under torch.compile

ydwu4 committed:
Update on "Automatically run dynamo in torch.cond"


The key modification is:
1.  moving cond from functorch/experimental/ to torch/_higher_order_ops/cond.py
2. always turn on dynamo in cond_compiled except on windows because torch.compile doesn't support windows rn.
3. change test behaviors after turning on torch.compile  for many eager tests.
4. expected fail a few tests in test_control_flow.py due to SymBool. Current plan is to wait for https://github.com/pytorch/pytorch/pull/107850 and forward fix.
5. Add new tests in test_control_flow.py to check cond can use dynamo to automatically lift the closures as input.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

ydwu4 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -82,7 +83,7 @@ def validate_args_and_maybe_create_graph_inputs(
                 example_value = a.as_proxy().node.meta["example_value"]
                 new_arg = wrap_fx_proxy(
-                    tx=tx, proxy=new_proxy, example_value=example_value
+                    tx=tx, proxy=new_proxy, example_value=example_value, source=a.source
```
Only inputs have source as I remember. It might be from some side-effects of some upstreamed diffs. I can check if we can remove it or not.

ydwu4 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -28,11 +30,51 @@
+@contextmanager
+def _set_compilation_env():
+    _old_is_tracing = torch.fx._symbolic_trace._is_fx_tracing_flag
+    _old_capture_dyn_ops = torch._dynamo.config.capture_dynamic_output_shape_ops
+    _old_capture_scalar = torch._dynamo.config.capture_scalar_outputs
```
For is_fx_tracing_flag, dynamo uses it to check whether we're in a tracin session [here](https://github.com/pytorch/pytorch/blob/71045f4885c1d8e684a0e4f2b64e2edf348c541a/torch/_dynamo/eval_frame.py#L311C29-L311C30). Right now, we have no problem in tracing make_fx( mode="real") so don't want to raise error on it since that's a lot of expectedFailures. We could consider remove the check in dynamo when things are ready.

For capture_dynamic_output_shape_ops and capture_scalar_outputs, they're used fro operations such as .nonzero and .item, we'd like to trace through them instead of having a graph break which will violate the fullgraph=True and a test failure.

zou3519 replied:
Can you point me to a test (or some sample code) where the `is_fx_tracing_flag` influences?

> For capture_dynamic_output_shape_ops and capture_scalar_outputs, they're used fro operations such as .nonzero and .item, we'd like to trace through them instead of having a graph break which will violate the fullgraph=True and a test failure.

Are there specific tests that you're looking at? Like I would have expected these to not work with regular torch.compile/export, and if so, these should error out with eager-mode cond.

ydwu4 replied:
> where the is_fx_tracing_flag influences

Yeah, basically all make_fx calls (despite the mode) will switch this flag to True. make_fx -> dispatch_trace -> tracer_base.trace -> https://github.com/pytorch/pytorch/blob/main/torch/fx/_symbolic_trace.py#L726 and have many tests in test_control_flow.py uses make_fx.

> Are there specific tests that you're looking at?

I couldn't find a concrete example but maybe we should try a bit on CI? It seems to me quite restrictive for cond not be able to call .item or nonzero inside. [export](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L1126) actually has a config to turn them on by default.

ydwu4 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -28,11 +30,51 @@
+    _old_acc_size_limit = torch._dynamo.config.accumulated_cache_size_limit
+
+    try:
+        # We need to turn off the is_fx_tracing_flag temporarily
+        # Remove it once we are confident fx tracing works with dynamo
+        torch.fx._symbolic_trace._is_fx_tracing_flag = False
```
Combined the answers in the above reply.

ydwu4 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -28,11 +30,51 @@
     reason: str
 
 
+def cond_compiled(pred, true_fn, false_fn, args):
```
Yeah, right now do you think it's OK to publish it despite some expected test failures? If you have no objection, I can do it in this diff.

ydwu4 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -28,11 +30,51 @@
 
 
+def cond_compiled(pred, true_fn, false_fn, args):
+    if torch._dynamo.is_compiling() or not torch._dynamo.is_dynamo_supported():
```
Sounds good!

ydwu4 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -28,11 +30,51 @@
 
+def cond_compiled(pred, true_fn, false_fn, args):
+    if torch._dynamo.is_compiling() or not torch._dynamo.is_dynamo_supported():
+        return cond(pred, true_fn, false_fn, args)
```
Yeah, that's a good point. It's not clear to me how to enforce it. In that case, should we just catch all the unimplemented error sin dynamo for cond and turn them into a different error that won't trigger dynamo re-compiliation so that dynamo won't fall back to eager?

ydwu4 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -28,11 +30,51 @@
+        return cond(pred, true_fn, false_fn, args)
+
+    with _set_compilation_env():
+        return torch.compile(wrapper, backend="eager", fullgraph=True)(
+            pred, true_fn, false_fn, args
+        )
```
we kind of rely on the automatic dynamic in dynamo and let guards do the job automatically so we don't actually care about whether it's dynamic or not in outer scope as long as they are backed symints  as described in https://github.com/pytorch/pytorch/pull/107662.

ydwu4 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -315,8 +359,8 @@ def cond_functionalize(interpreter, pred, true_fn, false_fn, inputs):
-cond.fallthrough(DispatchKey.PythonDispatcher)
-cond.fallthrough(DispatchKey.PythonTLSSnapshot)
+cond.fallthrough(DispatchKey.PythonDispatcher)  # type: ignore[attr-defined]
+cond.fallthrough(DispatchKey.PythonTLSSnapshot)  # type: ignore[attr-defined]
 cond.fallthrough(DispatchKey.ADInplaceOrView)
 cond.fallthrough(DispatchKey.BackendSelect)
-cond.fallthrough(DispatchKey.AutocastCPU)
+cond.fallthrough(DispatchKey.AutocastCPU)  # type: ignore[attr-defined]
```
OK!

ydwu4 committed:
Update on "Automatically run dynamo in torch.cond"


The key modification is:
1.  moving cond from functorch/experimental/ to torch/_higher_order_ops/cond.py
2. always turn on dynamo in cond_compiled except on windows because torch.compile doesn't support windows rn.
3. change test behaviors after turning on torch.compile  for many eager tests.
4. expected fail a few tests in test_control_flow.py due to SymBool. Current plan is to wait for https://github.com/pytorch/pytorch/pull/107850 and forward fix.
5. Add new tests in test_control_flow.py to check cond can use dynamo to automatically lift the closures as input.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

ydwu4 committed:
Update on "Automatically run dynamo in torch.cond"


The key modification is:
1.  moving cond from functorch/experimental/ to torch/_higher_order_ops/cond.py
2. always turn on dynamo in cond_compiled except on windows because torch.compile doesn't support windows rn.
3. change test behaviors after turning on torch.compile  for many eager tests.
4. expected fail a few tests in test_control_flow.py due to SymBool. Current plan is to wait for https://github.com/pytorch/pytorch/pull/107850 and forward fix.
5. Add new tests in test_control_flow.py to check cond can use dynamo to automatically lift the closures as input.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

ydwu4 committed:
Update on "Automatically run dynamo in torch.cond"


The key modification is:
1.  moving cond from functorch/experimental/ to torch/_higher_order_ops/cond.py
2. always turn on dynamo in cond_compiled except on windows because torch.compile doesn't support windows rn.
3. change test behaviors after turning on torch.compile  for many eager tests.
4. expected fail a few tests in test_control_flow.py due to SymBool. Current plan is to wait for https://github.com/pytorch/pytorch/pull/107850 and forward fix.
5. Add new tests in test_control_flow.py to check cond can use dynamo to automatically lift the closures as input.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

ydwu4 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -28,11 +30,51 @@
+        # We need to enlarge the dynamo cache size limit because
+        # cond's frame is expected to be hitted with different inputs frequently
+        torch._dynamo.config.cache_size_limit = 256
+        torch._dynamo.config.accumulated_cache_size_limit = 256
```
Each cond call is going to check against the cache, and since we didn't reset dynamo between tests, we end up accumulating a lot of caches for that cond operation.

May be we should mannuly call dynamo.reset() between the tests?

ydwu4 committed:
Update on "Automatically run dynamo in torch.cond"


The key modification is:
1.  moving cond from functorch/experimental/ to torch/_higher_order_ops/cond.py
2. always turn on dynamo in cond_compiled except on windows because torch.compile doesn't support windows rn.
3. change test behaviors after turning on torch.compile  for many eager tests.
4. expected fail a few tests in test_control_flow.py due to SymBool. Current plan is to wait for https://github.com/pytorch/pytorch/pull/107850 and forward fix.
5. Add new tests in test_control_flow.py to check cond can use dynamo to automatically lift the closures as input.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 committed:
Update on "Automatically run dynamo in torch.cond"


The key modification is:
1.  moving cond from functorch/experimental/ to torch/_higher_order_ops/cond.py
2. always turn on dynamo in cond_compiled except on windows because torch.compile doesn't support windows rn.
3. change test behaviors after turning on torch.compile  for many eager tests.
4. expected fail a few tests in test_control_flow.py due to SymBool. Current plan is to wait for https://github.com/pytorch/pytorch/pull/107850 and forward fix.
5. Add new tests in test_control_flow.py to check cond can use dynamo to automatically lift the closures as input.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

ydwu4 committed:
Update on "Automatically run dynamo in torch.cond"


The key modification is:
1.  moving cond from functorch/experimental/ to torch/_higher_order_ops/cond.py
2. always turn on dynamo in cond_compiled except on windows because torch.compile doesn't support windows rn.
3. change test behaviors after turning on torch.compile  for many eager tests.
4. expected fail a few tests in test_control_flow.py due to SymBool. Current plan is to wait for https://github.com/pytorch/pytorch/pull/107850 and forward fix.
5. Add new tests in test_control_flow.py to check cond can use dynamo to automatically lift the closures as input.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

zou3519 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -28,6 +30,30 @@
+        torch.fx._symbolic_trace._is_fx_tracing_flag = False
+        torch._dynamo.config.capture_dynamic_output_shape_ops = True
+        torch._dynamo.config.capture_scalar_outputs = True
+        # We need to enlarge the dynamo cache size limit because
+        # cond_wrapper's frame is expected to be hitted with different inputs frequently
+        torch._dynamo.config.cache_size_limit = 256
```
Is this just for testing? Or is there something else that we need to worry about?

ydwu4 replied:
I think it's only for testing. I don't expect users to call that many amount of times. One thing in my mind is that if the cahce_size_limit hits, dynamo will throw an unimplemented, which will be turned into an error in fullgraph mode. So we're probably safe.

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -351,13 +352,15 @@ def call_function(
         # branches
         assert isinstance(
-            args[1], (UserFunctionVariable, NestedUserFunctionVariable)
+            args[1],
+            (UserFunctionVariable, NestedUserFunctionVariable, NNModuleVariable),
```
If you want, splitting out the tuple change into a separate PR would make it easier to review

ydwu4 replied:
Sounds good!

zou3519 commented on torch/__init__.py:
```diff
@@ -1725,6 +1725,83 @@ def fn(model: Callable):
 
 from torch import export as export
 
+def cond(pred, true_branch, false_branch, operands):
```
We shouldn't put this directly into torch/__init__.py, there should be a `from torch._ops._higher_order.control_flow import cond` or something in `torch/__init__.py`.

To pacify our public API tests, we need to set cond's `.__module__` to "torch"

ydwu4 replied:
Ok! How to do the set cond's .__module__ to "torch"? I remember a exposed_in helper function in functorch. Can I use that?

zou3519 replied:
Yes, exposed_in

zou3519 commented on torch/__init__.py:
```diff
@@ -1725,6 +1725,83 @@ def fn(model: Callable):
+
+        - The **output** of branches must be a **single Tensor**. Pytree of tensors will be supported in the future.
+
+    """
```
I expect some input validation. That is:
- `pred` should be a bool, or a boolean Tensor
- true_fn/false_fn should be Callable
- operands must be Tuple. Furthermore, we require them to be Tuple[Tensor] right now.

ydwu4 replied:
Makes sense!

zou3519 replied:
NB: I think allowing List for torch.ops.higher_order.cond was a mistake. It's probably too late to do anything about that one. However, we can say that in our torch.cond API, we only want it to handle tuple operand.

zou3519 commented on test/functorch/test_control_flow.py:
```diff
@@ -248,6 +251,7 @@ def f(x, pred, pred2):
         graph = make_fx(f, tracing_mode="symbolic")(x, torch.tensor(False), torch.tensor(False))
         self.assertEqual(graph(x, torch.tensor(True), torch.tensor(True)), f(x, torch.tensor(True), torch.tensor(True)))
 
+    @unittest.expectedFailure
```
What's the problem with cond and functionalization?

ydwu4 replied:
It's due to the make_fx(tracing_mode="symbolic") call in L282. For eager functionalization and tracing_mode="real", they are both good.

zou3519 replied:
What's wrong with tracing_mode="symbolic"? The symbool thing? Or are there more issues?

ydwu4 replied:
Yeah, the symbol thing.

zou3519 commented on test/functorch/test_control_flow.py:
```diff
@@ -460,16 +465,20 @@ def f(x):
-            example_input_func = to_fun(example_input)
-            torch._enable_functionalization(reapply_views=False)
-            with self.assertRaisesRegex(UnsupportedAliasMutationException, "One of torch.cond branch"):
+        example_input_func = to_fun(example_input)
+        with self.assertRaisesRegex(UnsupportedAliasMutationException, "One of torch.cond branch"):
+            try:
+                torch._enable_functionalization(reapply_views=False)
```
Why did this change?

ydwu4 replied:
I remeber it is due to PYTORCH_TEST_WITH_DYNAMO=1 flag, let me check if i can remove the change.

zou3519 commented on test/functorch/test_control_flow.py:
```diff
@@ -796,8 +801,8 @@ def forward(self, x_1, pred_1, pred2_1):
-        def forward(self, y_1):
-            mul = torch.ops.aten.mul.Tensor(y_1, y_1);  y_1 = None
+        def forward(self, arg0_1):
+            mul = torch.ops.aten.mul.Tensor(arg0_1, arg0_1);  arg0_1 = None
```
(not blocking) I don't know why we're using self.assertEqual for these, they should all be self.assertExpectedInline

ydwu4 replied:
Can clean them up in next diff.

zou3519 commented on test/functorch/test_control_flow.py:
```diff
@@ -1210,7 +1216,104 @@ def foo(x):
+        inp = torch.ones(2, 3)
+        res = foo(inp)
+        # expected branches takes [x, a, b] as input
+        self._check_closure_correcly_lifted(foo, args=(inp, ), exp_res=res, exp_arg_num=3)
```
can you add an assertExpectedInline here for what make_fx returns? It doesn't need to be for all of your tests, just this one is fine

zou3519 commented on test/functorch/test_control_flow.py:
```diff
@@ -1210,7 +1216,104 @@ def foo(x):
+
+
+        inp = torch.ones(2, 3)
+        res = foo(inp)
```
I want to make sure that cond is still correct when a/b change. Could we add some tests where we do foo(inp), then modify a and b, and then call foo(inp) again?

zou3519 reviewed (commented):
My high-level comments are:
- I don't think we should expose torch.cond without the BACKEND_MATCH PR to make sure it is actually correct. I'm fine if you want to toss this into functorch.experimental.cond before that happens.
- How did we get Dynamo to error out on torch.ops.higher_order.cond if there are any graph breaks? (Can you add that to the PR description?)
- This PR might be easier to review if we split it up. For example, the change that adds tuple is its own fix, getting Dynamo to error out on torch.ops.higher_order.cond is another, one that introduces compiled_cond, and finally one at the end that turns the new compiled_cond function into torch.cond. It's a bit difficult right now to look at the expected failures and see if they are expected or not.
- Some of the config flag stuff is confusing
- Testing

On eager-mode cond specific Testing:
- eager-mode cond should handles captures nicely. I see tests for these with make_fx, that is good. We should also add tests to see that things are still correct if those captures change.
- eager-mode cond shouldn't execute any Tensor compute until we get after Dynamo. This is probably more difficult to test (I'm not really sure how right now)
- we expect eager-mode cond to work with make_fx and functionalization are called on it (I see some expected failures there, so I'm not sure what's up)

zou3519 commented on torch/__init__.py:
```diff
@@ -1725,6 +1725,83 @@ def fn(model: Callable):
+
+    if torch._dynamo.is_compiling():
+        return torch.ops.higher_order.cond(pred, true_branch, false_branch, operands)
+
```
One thing on true_fn/false_fn validation: Our restriction is that true_fn/false_fn's inputs must not be mutated, and we must not return a view of the input. This is important to get right for BC reasons. IIRC this validation is in the functionalization code, which doesn't actually get run... maybe we would need to move it into dynamo handling for cond.

ydwu4 committed:
Update on "Automatically run dynamo in torch.cond"


The key modification is:
1.  moving cond from functorch/experimental/ to torch/_higher_order_ops/cond.py
2. always turn on dynamo in cond_compiled except on windows because torch.compile doesn't support windows rn.
3. change test behaviors after turning on torch.compile  for many eager tests.
4. expected fail a few tests in test_control_flow.py due to SymBool. Current plan is to wait for https://github.com/pytorch/pytorch/pull/107850 and forward fix.
5. Add new tests in test_control_flow.py to check cond can use dynamo to automatically lift the closures as input.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

ydwu4 committed:
Update on "Automatically run dynamo in torch.cond"


The key modification is:
1.  moving cond from functorch/experimental/ to torch/_higher_order_ops/cond.py
2. always turn on dynamo in cond_compiled except on windows because torch.compile doesn't support windows rn.
3. change test behaviors after turning on torch.compile  for many eager tests.
4. expected fail a few tests in test_control_flow.py due to SymBool. Current plan is to wait for https://github.com/pytorch/pytorch/pull/107850 and forward fix.
5. Add new tests in test_control_flow.py to check cond can use dynamo to automatically lift the closures as input.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

ydwu4 commented on test/dynamo/test_logging.py:
```diff
@@ -388,10 +388,10 @@ def outer(pred, x):
             if "return x * 2" in msg:
                 found_x2 = True
-                self.assertIn("inline depth: 2", msg)
+                self.assertIn("inline depth: 3", msg)
```
torch._higher_order_ops.cond.cond  is treated as a UserFunctionVariable by dynamo. so it increases the inline depth by one

ydwu4 committed:
Update on "Automatically run dynamo in torch.cond"


The key modification is:
1.  moving cond from functorch/experimental/ to torch/_higher_order_ops/cond.py
2. always turn on dynamo in cond_compiled except on windows because torch.compile doesn't support windows rn.
3. change test behaviors after turning on torch.compile  for many eager tests.
4. expected fail a few tests in test_control_flow.py due to SymBool. Current plan is to wait for https://github.com/pytorch/pytorch/pull/107850 and forward fix.
5. Add new tests in test_control_flow.py to check cond can use dynamo to automatically lift the closures as input.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

ydwu4 committed:
Update on "Automatically run dynamo in torch.cond"


The key modification is:
1.  moving cond from functorch/experimental/ to torch/_higher_order_ops/cond.py
2. always turn on dynamo in cond_compiled except on windows because torch.compile doesn't support windows rn.
3. change test behaviors after turning on torch.compile  for many eager tests.
4. expected fail a few tests in test_control_flow.py due to SymBool. Current plan is to wait for https://github.com/pytorch/pytorch/pull/107850 and forward fix.
5. Add new tests in test_control_flow.py to check cond can use dynamo to automatically lift the closures as input.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

zou3519 commented on functorch/experimental/_map.py:
```diff
@@ -74,9 +74,15 @@ def create_fw_bw_graph(f, num_mapped_args, *args):
-                    return torch.empty_strided(
-                        t.size(), t.stride(), requires_grad=t.requires_grad
-                    )
+                    if t.dtype != torch.bool:
```
What's going on here?

ydwu4 replied:
boolean tensor created by empty_strided is not properly intialized so it causes error "load of value 190, which is not a valid value for type 'bool'" on clang asan environment on CI.

ydwu4 commented:
replaced by #108028
